libraries/rush-lib/src/logic/snapshots/InputSnapshot.ts

```diff
@@ -0,0 +1,437 @@
+   * @returns A map of file name to Git hash. For local files paths will be relative. Configured additional files may be absolute paths.
+   */
+  getTrackedFileHashesForOperation(
+    project: IRushConfigurationProjectForSnapshot,
+    operationName?: string
+  ): ReadonlyMap<string, string>;
+
+  /**
+   * Gets the local state hash for the operation. This will later be combined with the hash of the command being executed and the final hashes of the operation's dependencies to compute
+   * the final hash for the operation.
+   * @param project - The Rush project to compute the state hash for
+   * @param operationName - The name of the operation (phase) to get hashes for. If omitted, returns a generic hash for the whole project, as used for bulk commands.
+   * @returns The local state hash for the project. This is a hash of the environment, the project's tracked files, and any additional files.
+   */
+  getLocalStateHashForOperation(
+    project: IRushConfigurationProjectForSnapshot,
+    operationName?: string
+  ): string;
+}
+
+/**
+ * Represents a synchronously-queryable in-memory snapshot of the state of the inputs to a Rush repository.
+ * Any asynchronous work needs to be performed by the caller and the results passed to the constructor.
+ *
+ * @remarks
+ * All operations on this class will return the same result regardless of when they are executed.
+ *
+ * @internal
+ */
+export class InputSnapshot implements IInputSnapshot {
+  /**
+   * The metadata for each project. This is a superset of the information in `projectMap` and includes caching of queries.
+   */
+  private readonly _projectMetadataMap: Map<
+    IRushConfigurationProjectForSnapshot,
+    IInternalRushSnapshotProjectMetadata
+  >;
+  /**
+   * Hashes of files to be included in all result sets.
+   */
+  private readonly _globalAdditionalHashes: ReadonlyMap<string, string> | undefined;
+  /**
+   * Hashes for files selected by `dependsOnAdditionalFiles`.
+   */
+  private readonly _additionalHashes: ReadonlyMap<string, string> | undefined;
+  /**
+   * The environment to use for `dependsOnEnvVars`.
+   */
+  private readonly _environment: Record<string, string | undefined>;
+  /**
+   * The hashes of all tracked files in the repository.
+   */
+  private readonly _hashes: ReadonlyMap<string, string>;
+
+  /**
+   *
+   * @param params - The parameters for the snapshot
+   * @internal
+   */
+  public constructor(params: IRushSnapshotParameters) {
+    const {
+      additionalHashes,
+      environment = { ...process.env },
+      globalAdditionalFiles,
+      hashes,
+      lookupByPath,
+      rootDir
+    } = params;
+    const projectMetadataMap: Map<
+      IRushConfigurationProjectForSnapshot,
+      IInternalRushSnapshotProjectMetadata
+    > = new Map();
+    for (const [project, record] of params.projectMap) {
```

Comment:
Is this just the projects that the command invocation will actually run operations on?